aioxmpp/ibr/service.py

```diff
@@ -0,0 +1,253 @@
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+########################################################################
+
+import asyncio
+
+import aioxmpp
+import logging
+
+from . import xso
+
+
+logger = logging.getLogger(__name__)
+
+
+class RegistrationService:
+    """
+    Service implementing XMPP In-Band Registration(:xep:`0077`).
+
+    This service implements the response to XMPP Pings and provides a method
+    to send pings.
+
+    This class doesn't inherit from the Service class because here are methods
+    that can't have a connected client in order to run. For example, a new
+    account should sent the needed stanza without being authenticated,
+    otherwise it would make no sense.
+
+    .. automethod:: get_client_info
+
+    .. automethod:: change_pass
+
+    .. automethod:: cancel_registration
+
+    .. automethod:: get_registration_fields
+
+    .. automethod:: register
+
+    """
+
+    @asyncio.coroutine
+    def get_client_info(self, client):
+        """
+        A query is sent to the server o obtain the client's data stored at the
+        server.
+
+        :param client: Client from which the query is sent.
+        :type client: :class:`aioxmpp.Client`
+
+        :return: IQ response :attr:`~.IQ.payload` or :data:`None`
+        """
+        iq = aioxmpp.IQ(
+            to=client.local_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.GET,
+            payload=xso.Query()
+        )
+
+        return (yield from client.send(iq))
+
+    @asyncio.coroutine
+    def change_pass(self,
+                    client,
+                    new_pass,
+                    aux_fields=None):
+        """
+        Change the client password for 'new_pass'.
+
+        :param client: Client who wants to change the password.
+        :type client: :class:`aioxmpp.Client`
+
+        :param new_pass: New password of the client.
+        :type new_pass: :class:`str`
+
+        :param aux_fields: Auxiliary fields in case additional info is needed.
+        :type aux_fields: :class:`dict`
+
+        :return:`None`
+
+        If the query fails for whatever reason, it will raise an exception
+        from inside client.send(iq).
+        """
+        iq = aioxmpp.IQ(
+            to=client.local_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.SET,
+            payload=xso.Query()
+        )
+
+        iq.payload.username = client.local_jid.localpart
+        iq.payload.password = new_pass
+
+        if aux_fields is not None:
+            for key, value in aux_fields.items():
+                setattr(iq.payload, key, value)
```

Comment:
I don’t think this is needed. XEP-0077 is rather clear that (in the non-data-form flow) only ``username`` and ``password`` need to be given to change the password. Let’s remove this complexity here for now.